### PR TITLE
Enable Scale and HPA support for VMIRS

### PR DESCRIPTION
**Pull request by rmohr**
**What this PR does / why we need it**:

Enable and implement the scale subresouce for VMIRS instances. This allows the following:

First, running `kubectl scale` commands against VMIRS. For instance:

```bash
kubectl scale vmirs myvmirs --replicas 5
```

This requires the `CustomResourceSubresources` feature gate to be enables for clusters prior to 1.11.

Second, allow using the Horizontal Pod Autoscaler. For instance:

```yaml
apiVersion: autoscaling/v1
kind: HorizontalPodAutoscaler
metadata:
  name: myhpa
spec:
  scaleTargetRef:
    kind: VirtualMachineInstanceReplicaSet
    name: vmi-replicaset-cirros
    apiVersion: kubevirt.io/v1alpha2
  minReplicas: 3
  maxReplicas: 10
  targetCPUUtilizationPercentage: 50
```

In combination with the HPA, there are two restrictions:

 * HPA works with CRDs starting from 1.11, tests are skipped for older cluster versions.
 * `kubectl autoscale` does not yet work with CRDs. Posting the HPA spec directly with `kubectl create` is necessary.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

Documentation is at https://github.com/kubevirt/user-guide/pull/190.

**Release note**:

```release-note
Enable scale and HPA support for VMIRS
```

### CODE REVIEW

**Comment by cynepco3hahue** [comment]
@rmohr I see failure under https://jenkins.ovirt.org/job/kubevirt_kubevirt_standard-check-pr/3003//artifact/check-patch.k8s-1.10.11.el7.x86_64/mock_logs/script/stdout_stderr.log
```
[Fail] VirtualMachineInstanceReplicaSet [It] should not scale when paused and scale when resume 
/root/go/src/kubevirt.io/kubevirt/tests/replicaset_test.go:355
```

**Comment by rmohr** [comment]
@cynepco3hahue thanks. Addressed. I use now `PATCH` where appropriate to avoid such collisions.

File: tests/replicaset_test.go [hunk 1]
```diff
@@ -29,17 +29,19 @@ import (
...
 	"k8s.io/apimachinery/pkg/util/json"
 
-	v1 "kubevirt.io/kubevirt/pkg/api/v1"
+	"kubevirt.io/kubevirt/pkg/api/v1"
 	"kubevirt.io/kubevirt/pkg/kubecli"
 	"kubevirt.io/kubevirt/tests"
 )
 
-var _ = Describe("VirtualMachineInstanceReplicaSet", func() {
+var _ = FDescribe("VirtualMachineInstanceReplicaSet", func() {
```
Comment (cynepco3hahue) [inline tests/replicaset_test.go]: Please remove `F`

File: pkg/kubecli/replicaset.go [hunk 2]
```diff
@@ -37,6 +39,31 @@ type rc struct {
...
+		Namespace(v.namespace).
+		Resource(v.resource).
+		Name(statefulSetName).
+		SubResource("scale").
+		Do().
+		Into(result)
+	return
+}
+
+func (v *rc) UpdateScale(statefulSetName string, scale *autov1.Scale) (result *autov1.Scale, err error) {
```
Comment (davidvossel) [inline pkg/kubecli/replicaset.go]: run this regex on the file. 's/statefulSetName/replicaSetName'

File: api/openapi-spec/swagger.json [hunk 3]
```diff
@@ -5893,6 +5893,10 @@
        "$ref": "#/definitions/v1.VirtualMachineInstanceReplicaSetCondition"
       }
      },
+     "labelSelector": {
```
Comment (davidvossel) [inline api/openapi-spec/swagger.json]: How is the HPA detecting load to know when to scale? Is it using this labelSelector and calculating load based on the Pods that match this selector?

File: manifests/generated/vmirs-resource.yaml [hunk 4]
```diff
@@ -32,4 +32,9 @@ spec:
     - vmirss
     singular: virtualmachineinstancereplicaset
   scope: Namespaced
+  subresources:
```
Comment (davidvossel) [inline manifests/generated/vmirs-resource.yaml]: > This requires the CustomResourceSubresources feature gate to be enables for clusters prior to 1.11.

Is this CRD going to get rejected for 1.10 if the feature gate isn't enabled?

File: api/openapi-spec/swagger.json (hunk 3 above)
Comment (rmohr) [reply api/openapi-spec/swagger.json]: Yes. The metrics api needs to be active inside the cluster: https://kubernetes.io/docs/tasks/debug-application-cluster/core-metrics-pipeline/.

We will have to improve the situation a little bit where no readiness check is explicitly configured on the VMI and migrations come into play, but apart from that the algorithm can work with non-ready pods, which allows to cope with "boot" load.

File: manifests/generated/vmirs-resource.yaml (hunk 4 above)
Comment (rmohr) [reply manifests/generated/vmirs-resource.yaml]: No, will just strip the fields away from the CRD. I find that actually a little bit unpleasant in general but good in this case ...

**Comment by rmohr** [comment]
Ok test results look good and I opened https://github.com/kubernetes/kubernetes/pull/72678 to add autoscale support for kubectl. Could someone have another look?

File: pkg/kubecli/replicaset.go (hunk 2 above)
Comment (rmohr) [reply pkg/kubecli/replicaset.go]: Done.

**Comment by cynepco3hahue** [comment]
lgtm

File: api/openapi-spec/swagger.json (hunk 3 above)
Comment (davidvossel) [reply api/openapi-spec/swagger.json]: yeah, batch migrations may confuse HPA. The workload's calculated metrics during a bulk migrations might not accurately reflect the actual workload due to migration overhead.

**Pull request merged**